compiler/passes/insertLineNumbers.cpp

```diff
@@ -77,356 +78,393 @@ int getFilenameLookupPosition(std::string name) {
-        call->isPrimitive(PRIM_TESTCID)          ||
-        isClassMethodCall(call)) {
-      Expr*          arg0 = call->get(1);
-      AggregateType* ct   = toAggregateType(arg0->typeInfo());
-
-      if (ct && (isClass(ct) || ct->symbol->hasFlag(FLAG_WIDE_CLASS))) {
-        FnSymbol* fn = call->resolvedFunction();
-
-        // Avoid inserting a nil-check if this is a call to a destructor
-        if (fn == NULL || fn->hasFlag(FLAG_DESTRUCTOR) == false) {
-          Expr* stmt = call->getStmtExpr();
+        call->isPrimitive(PRIM_SET_MEMBER) ||
+        call->isPrimitive(PRIM_GETCID) ||
+        call->isPrimitive(PRIM_TESTCID) ||
+        isClassMethodCall(call))) {
+    return false;
+  }
 
-          SET_LINENO(stmt);
+  Expr* arg0 = call->get(1);
+  AggregateType* ct = toAggregateType(arg0->typeInfo());
 
-          stmt->insertBefore(new CallExpr(PRIM_CHECK_NIL, arg0->copy()));
-        }
-      }
+  if (ct && (isClass(ct) || ct->symbol->hasFlag(FLAG_WIDE_CLASS))) {
+    // Avoid inserting a nil-check if this is a call to a destructor
+    if (FnSymbol* fn = call->resolvedFunction()) {
+      return !fn->hasFlag(FLAG_DESTRUCTOR);
     }
+
+    return true;
   }
+
+  return false;
 }
 
-void insertLineNumbers() {
-  compute_call_sites();
+void InsertNilChecksPass::process(CallExpr* call) {
+  Expr* arg0 = call->get(1);
+  Expr* stmt = call->getStmtExpr();
 
-  // Temporary vector that stores some constant filenames that are used
-  // directly by the runtime. The index for these matter, and are provided to
-  // the runtime as defines in chpl-linefile-support.h
-  std::vector<std::string> constantFilenames;
+  SET_LINENO(stmt);
 
-  // Put a null string into the iterator at the first position, some runtime
-  // calls will pass in NULL for their filename, we can then use this null
-  // string to deal with that case.
-  constantFilenames.push_back("");
+  stmt->insertBefore(new CallExpr(PRIM_CHECK_NIL, arg0->copy()));
+}
 
-  // Add "<internal>" to the filename table if it didn't make it in there, some
-  // runtime functions use this name directly, and it doesn't always end up in
-  // the table otherwise
-  constantFilenames.push_back("<internal>");
+// --------------------
 
-  gFilenameLookup.insert(gFilenameLookup.begin(),
-                         constantFilenames.begin(),
-                         constantFilenames.end());
+#include "global-ast-vecs.h"
 
-  if (!fNoNilChecks) {
-    insertNilChecks();
-  }
+void insertLineNumbers() {
+  PassManager pm;
 
-  // loop over all primitives that require a line number and filename
-  // and pass them an actual line number and filename
-  forv_Vec(CallExpr, call, gCallExprs) {
-    if (call->primitive && call->primitive->passLineno) {
-      insertLineNumber(call);
-    }
-  }
+  // It's pretty apparent in this refactoring that we don't actually need to
+  // recompute all call sites, but only the subset we'll be working on, which
+  // is probably better done on demand
```

Comment:
I think this comment is mentioning a TODO. Could you prefix it with TODO to make that easier to understand?

Reply:
added